clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1875,55 +1873,93 @@ namespace {
             S->dump(llvm::outs());
             llvm::outs().flush();
 #endif
-            TraverseStmt(S, CSS);
+            Check(S);
          }
        }
        AFA.Next();
      }
     }
 
-    // Traverse methods iterate recursively over AST tree nodes, visiting
-    // all children of any node not visited by a Check method.
+  // Methods for inferring bounds expressions for C expressions.
+
+  // C has an interesting semantics for expressions that differentiates between
+  // lvalue and value expressions and inserts implicit conversions from lvalues
+  // to values.  Value expressions are usually called rvalue expressions.  This
+  // semantics is represented directly in the clang IR by having some
+  // expressions evaluate to lvalues and having implicit conversions that convert
+  // those lvalues to rvalues.
+  //
+  // Using ths representation directly would make it clumsy to compute bounds
+  // expressions.  For an expression that evaluates to an lvalue, we would have
+  // to compute and carry along two bounds expressions: the bounds expression
+  // for the lvalue and the bounds expression for the value at which the lvalue
+  // points.
+  //
+  // We address this by having two methods for computing bounds.  One method
+  // (Check) computes the bounds for an rvalue expression. For lvalue
+  // expressions, we have one method that compute two kinds of bounds.
+  // CheckLValue computes the bounds for the lvalue produced by an expression
+  // and the bounds for the target of the lvalue produced by the expression.
+  //
+  // There are only a few contexts where an lvalue expression can occur, so it
+  // is straightforward to determine which method to use. Also, the clang IR
+  // makes it explicit when an lvalue is converted to an rvalue by an lvalue
+  // cast operation.
+  //
+  // An expression denotes an lvalue if it occurs in the following contexts:
+  // 1. As the left-hand side of an assignment operator.
+  // 2. As the operand to a postfix or prefix incrementation operators (which
+  //    implicitly do assignment).
+  // 3. As the operand of the address-of (&) operator.
+  // 4. If a member access operation e1.f denotes an lvalue, e1 denotes an
+  //    lvalue.
+  // 5. In clang IR, as an operand to an LValueToRValue cast operation.
+  // Otherwise an expression denotes an rvalue.
+
+  public:
+    // If e is an rvalue, Check returns the bounds for the value produced by e.
```

Comment:
I think this comment still needs to be updated.  Checked checks e and its children, even if it is an rvalue.